qemu/target/riscv/unicorn.c

```diff
@@ -79,6 +79,63 @@ static void riscv_release(void *ctx)
 
 static void reg_reset(struct uc_struct *uc) {}
 
+static uc_err reg_read_priv(CPURISCVState *env, target_ulong *value)
+{
+    // This structure is based on RISC-V Debug Specification 1.0.0-rc3,
+    // Section 4.10.1, Virtual Debug Registers: Privilege Mode.
+    // This encoding should match the decoding in reg_write_priv.
+    target_ulong priv_value = 0;
+    switch (env->priv) {
+    default:
+        // No other value should be possible, but we'll report
+        // 0 (U-Mode) in this case since that's most conservative.
+        break;
+    case PRV_U:
+        priv_value = 0;
+        break;
+    case PRV_S:
+        priv_value = 1;
+        break;
+    case PRV_M:
+        priv_value = 3;
+        break;
+    }
+    if (riscv_cpu_virt_enabled(env)) {
+        // The "v" bit is set to indicate either VS or VU mode.
+        priv_value |= 0b100;
+    }
+    *value = priv_value;
+    return UC_ERR_OK;
+}
+
+static uc_err reg_write_priv(CPURISCVState *env, target_ulong value)
+{
+    // This structure is based on RISC-V Debug Specification 1.0.0-rc3,
+    // Section 4.10.1, Virtual Debug Registers: Privilege Mode.
+    // This decoding should match the encoding in reg_read_priv.
+    if ((value & ~0b111) != 0) {
+        // Only the low three bits are settable.
+        return UC_ERR_ARG;
+    }
+    target_ulong prv = value & 0b11;
+    bool v = (value & 0b100) != 0;
+    switch (prv) {
```

Comment:
The `PRV_U`, `PRV_S`, and `PRV_M` constants do happen to match this encoding and so I originally had this written to just accept values directly, but I decided to write this out more explicitly because it seemed unwise to expose the assumption that QEMU's `priv` field will always match the debug spec's encoding in future versions, and also unwise to expose the possibility of setting the reserved value 2 that might take on a different meaning in a future QEMU version.